### ISSUE TITLE
Another signal 11 for recursive generic type declaration

### ISSUE THREAD

**Issue by BrucePerens** [crystal-lang/crystal#9936]
This gives a signal 11:
```crystal
alias AllowedValue = Bool|Allowed
alias Allowed = Hash(Symbol, Bool|Hash(Symbol, AllowedValue))
h : Allowed
a = { :symbol => true } of Symbol => Bool|Allowed
h = { :symbol => a }    of Symbol => Bool|Allowed
```

**Comment by asterite** [comment]
I think this is essentially #3804

**Comment by straight-shoota** [comment]
Sadly, we haven't managed to remove recursive aliases for three years now (#5155) 😢 

I'd recommend everyone to not use them. They're unnecessary and the implementation broken in many ways.

**Comment by BrucePerens** [comment]
@straight-shoota if you don't want people to use recursive aliases, `stdlib` should contain an alternative that they can be pointed to. As a relative Crystal neophyte this is the best I could do this evening, surely you can do better:
```crystal
# Implementation of a pseudo-generic that contains itself without use of
# recursively-defined aliases, which are problematical and/or broken
# in the compiler.
macro recursive_hash(name, keytype, valuetype)
  struct Type_%a
    property value : {{valuetype.id}}
    def initialize(@value)
    end
  end
  
  class {{name.id}}
    @h = Hash({{keytype.id}}, Type_%a).new

    def []=(key, value)
      @h[key] = Type_%a.new(value)
    end
  end
end

recursive_hash(MyHash, Symbol, String|MyHash|Array(MyHash))

h = MyHash.new
h[:itself] = h
h[:array] = [MyHash.new]
p h.inspect
```

**Comment by asterite** [comment]
@BrucePerens What's your specific use case here?

**Comment by BrucePerens** [comment]
@asterite The use case is an interpretive language. It has arrays which contain its programs and all of their data types, potentially including hashes, and hashes that contain its data (and optionally more hashes). I coded it using aliases for self-containing generics: Hash and Array. The language is potentially multi-threaded and I will probably eventually add Mutex to these containers, as I am dubious about container write being atomic with regard to a simultaneous read.

There were some bumps on the way, including the compiler getting a signal 11 without emitting any debugging information that would help me to understand why (a signal catcher that emits an internal stack dump might be helpful, I've not yet attempted running lldb on the compiler and the multiple threads will be a complication if I start).

**Comment by asterite** [comment]
I see.

Yes, in that case making your own data type (something like Ruby's VALUE in C) is probably better than using recursive alias, specially because of how buggy its implementation is.

**Comment by BrucePerens** [comment]
Yes. It is annoying that the cost of this is that I need to implement a pass-through method for every method of Hash and Array. At least I need only do it once, and I'll package the macros for others.

While recursively-typed aliases might not be the solution, it would be nice for there to be a lot of repair on generics, and for there to be broader rules on the use of the generic type arguments - which are not currently in scope when used for much other than the right side of `variable :` or as arguments to another generic. I would also like for generics to take a type keyword like `itself` so that I could instantiate them this way: `Hash(Symbol, String|itself|Array(itself))`.

Thanks

Bruce

**Comment by straight-shoota** [comment]
@BrucePerens You might take a look at https://github.com/straight-shoota/crinja/blob/7511403d058d6dc505ced2344a78491bd0fe9cc3/src/runtime/value.cr Seems like a similar role to what you're trying to do.

**Comment by BrucePerens** [comment]
@straight-shoota OK, I will do that. I see you use `tap` and I have to get my head around that - I have not had much clue what it's for up until now.

**Comment by Sija** [comment]
@BrucePerens Also, you can check [any_hash](https://github.com/Sija/any_hash.cr) shard.

**Comment by BrucePerens** [comment]
@Sija Thank you, there are lots of interesting lessons I need to learn in there.

**Comment by BrucePerens** [comment]
Initial code at https://github.com/BrucePerens/recursive_generic . Not ready for you to use.
